apis/Google.Cloud.Metadata.V1/Google.Cloud.Metadata.V1/MetadataClientImpl.cs

```diff
@@ -0,0 +1,502 @@
+        {
+            switch (result)
+            {
+                case "NONE": return MaintenanceStatus.None;
+                case "MIGRATE_ON_HOST_MAINTENANCE": return MaintenanceStatus.Migrate;
+                default: return MaintenanceStatus.Unknown;
+            }
+        }
+
+        private string GetMetadataHost()
+        {
+            var emulatorHost = Environment.GetEnvironmentVariable(EmulatorEnvironmentVariable);
+            return string.IsNullOrEmpty(emulatorHost) ? DefaultMetadataHost : emulatorHost;
+        }
+
+        private static T GetResult<T>(Func<CancellationToken, Task<T>> operation)
+        {
+            try
+            {
+                return operation(CancellationToken.None).Result;
+            }
+            catch (AggregateException e)
+            {
+                throw e.InnerExceptions.FirstOrDefault() ?? e;
+            }
+        }
+
+        private async Task<HttpResponseMessage> RequestMetadata(string relativeUrl, CancellationToken cancellationToken)
+        {
+            var httpRequest = new HttpRequestMessage(HttpMethod.Get, $"http://{metadataHost}/computeMetadata/v1/{relativeUrl}");
+            httpRequest.Headers.Add(MetadataFlavor, GoogleMetadataHeader);
+            return await HttpClient.SendAsync(httpRequest, cancellationToken).ConfigureAwait(false);
+        }
+
+        private static readonly Regex KeyRegex = new Regex(@"\A[a-zA-Z0-9-_]{1,127}\Z", RegexOptions.Compiled);
+        private static void ValidateCustomMetadataKey(string key)
+        {
+            if (!KeyRegex.IsMatch(key))
+            {
+                throw new ArgumentException("Invalid custom metadata key.", "key");
+            }
+        }
+
+        /// <inheritdoc/>
+        public override TokenResponse GetAccessToken() => GetResult(GetAccessTokenAsync);
+
+        /// <inheritdoc/>
+        public override async Task<TokenResponse> GetAccessTokenAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            // TODO: Most of this is taken from Google.Apis.Auth.OAuth2.ComputeCredential.RequestAccessTokenAsync. Perhaps move that to a helper which can be shared.
+            var response = await RequestMetadata("instance/service-accounts/default/token", cancellationToken).ConfigureAwait(false);
+
+            // Read the response.
+            var content = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+            if (!response.IsSuccessStatusCode)
+            {
+                try
+                {
+                    var error = NewtonsoftJsonSerializer.Instance.Deserialize<TokenErrorResponse>(content);
+                    throw new TokenResponseException(error);
+                }
+                // Json.NET deserialize object method doesn't include documentation about which exceptions can
+                // be thrown, that's why we just catch a general exception here.
+                catch (Exception ex)
```

Comment:
You can catch `JsonException` instead, see [here](http://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_JsonException.htm)